### PR TITLE
Add a Jenkinsfile to build, test, and tag the Dockerfile

### PR DESCRIPTION
**Pull request by bigspotteddog**
https://issues.sonatype.org/browse/INT-352
https://jenkins.zion.aws.s/job/integrations/job/docker-nexus-iq-server-feature/job/INT-332_jenkins_build_and_tag

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      OsTools.runSafe(this, "git tag -d ${version}")
+    }
+  } finally {
+    OsTools.runSafe(this, 'git clean -f && git reset --hard origin/master')
+  }
+}
+def readVersion() {
+  def content = readFile 'Dockerfile'
+  for (line in content.split('\n')) {
+    if (line.startsWith('ENV IQ_SERVER_COOKBOOK_VERSION="release-')) {
```
Comment (bigspotteddog) [inline Jenkinsfile]: Need a better way to determine the version for this release. Semantic? Or, the nexus-iq-server release as sonatype-nexus3 is done?

Comment (bigspotteddog) [reply Jenkinsfile]: Used the IQ_SERVER_VERSION version that is now set in the Dockerfile (89d8567).

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                        usernameVariable: 'GITHUB_API_USERNAME', passwordVariable: 'GITHUB_API_PASSWORD']]) {
+        apiToken = env.GITHUB_API_PASSWORD
+      }
+      gitHub = new GitHub(this, "${gitHubUsername}/${gitHubRepository}", apiToken)
+    }
+    stage('Build') {
+      gitHub.statusUpdate commitId, 'pending', 'build', 'Build is running'
+
+      def gemInstallDirectory = getGemInstallDirectory()
+      withEnv(["PATH+GEMS=${gemInstallDirectory}/bin"]) {
```
Comment (fernau) [inline Jenkinsfile]: /

File: Dockerfile [hunk 3]
```diff
@@ -21,8 +21,8 @@ LABEL vendor=Sonatype \
   com.sonatype.name="Nexus IQ Server image"
 
 # Optional parameters. Uncomment to override default:
-# ENV IQ_SERVER_VERSION=""
-# ENV IQ_SERVER_SHA256=""
+ENV IQ_SERVER_VERSION=1.36.0-01
+ENV IQ_SERVER_SHA256=a1a5d127e2deaf2b26ed4e605802986b99ea317c8a821059c918783517728d6c
```
Comment (fernau) [inline Dockerfile:25]: are these parameters now mandatory or is this just for testing ?

**Review by fernau — APPROVED** [review]
+1

File: Dockerfile (hunk 3 above)
Comment (bigspotteddog) [reply Dockerfile:25]: I was thinking these would be mandatory tagging purposes, similar to how the Nexus Repository Manager docker image is done.

Another idea would be for the chef recipe to somehow communicate the version it is building; maybe writing to a file.

Comment (fernau) [reply Dockerfile:25]: no it's better to leave it as it is was just asking. thanks

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import com.sonatype.jenkins.pipeline.GitHub
+import com.sonatype.jenkins.pipeline.OsTools
+
+node('ubuntu-zion') {
+  def commitId, commitDate, version, gitHubUsername, gitHubRepository, credentialsId, imageName, archiveName
+  GitHub gitHub
+
+  try {
+    stage('Preparation') {
+      gitHubUsername = 'sonatype'
```
Comment (whyjustin) [inline Jenkinsfile]: This is an organization, not a user

Comment (whyjustin) [reply Jenkinsfile]: These can be declared when they are defined.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      archiveName = 'sonatype-nexus-iq-server'
+
+      deleteDir()
+
+      checkout scm
+
+      commitId = OsTools.runSafe(this, 'git rev-parse HEAD')
+      commitDate = OsTools.runSafe(this, "git show -s --format=%cd --date=format:%Y%m%d-%H%M%S ${commitId}")
+
+      buildType = scm.branches[0].name != '*/master' ? 'snapshot' : 'release'
```
Comment (whyjustin) [inline Jenkinsfile]: This isn't used

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    stage('Build') {
+      gitHub.statusUpdate commitId, 'pending', 'build', 'Build is running'
+
+      def gemInstallDirectory = getGemInstallDirectory()
+      withEnv(["PATH+GEMS=${gemInstallDirectory}/bin"]) {
+        OsTools.runSafe(this, "docker system prune -a -f")
+        OsTools.runSafe(this, "gem install --user-install rspec")
+        OsTools.runSafe(this, "gem install --user-install serverspec")
+        OsTools.runSafe(this, "gem install --user-install docker-api")
+        OsTools.runSafe(this, "rspec --backtrace spec/Dockerfile_spec.rb")
```
Comment (whyjustin) [inline Jenkinsfile]: We should build the image in a separate stage and then test the built image, separating the build from the testing.

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        gitHub.statusUpdate commitId, 'success', 'build', 'Build succeeded'
+      }
+    }
+    stage('Archive') {
+      dir('build/target') {
+        OsTools.runSafe(this, "docker save ${imageName} | gzip > ${archiveName}.tar.gz")
+        archiveArtifacts artifacts: "${archiveName}.tar.gz", onlyIfSuccessful: true
+      }
+    }
+    if (currentBuild.result == 'FAILURE') {
```
Comment (whyjustin) [inline Jenkinsfile]: We don't need to archive on failure

**Review by whyjustin — CHANGES_REQUESTED** [review]
Some feedback - needs change to push to Docker Hub as well.

File: Jenkinsfile (hunk 4 above)
Comment (bigspotteddog) [reply Jenkinsfile]: Declared when defined (429c642).

File: Jenkinsfile (hunk 7 above)
Comment (bigspotteddog) [reply Jenkinsfile]: Moved the archive below the failure check (429c642).

File: Jenkinsfile (hunk 5 above)
Comment (bigspotteddog) [reply Jenkinsfile]: Removed this (429c642).

**Comment by bigspotteddog** [comment]
Applied the same comments to https://github.com/sonatype/docker-nexus3/pull/81.

File: Jenkinsfile (hunk 6 above)
Comment (bigspotteddog) [reply Jenkinsfile]: Build first, then test (ba15074).

**Review by whyjustin — APPROVED** [review]

**Pull request merged**